smolyay/samples.py

```diff
@@ -0,0 +1,207 @@
+import abc
+import sys
+import numpy
+
+
+class UnidimensionalPointSet(abc.ABC):
```

Comment:
Should all `UnidimensionalPointSet`s be given a concept of a level the points are assigned at? For sets that don't really have levels, they can all be dumped into the first level.

Reply:
It depends on whether all the MultidimensionPointSets that use a UnidimensionalPointSet will care about levels. A smolyak point set will care, but a tensor point set doesn't have to.

Reply:
Fair enough. I am good with any of the following options:

1. Force all `UnidimensionalPointSet`s to assign their points to a level (and perhaps default it to `None` if there isn't a concept of a level for a certain set).
2. Create a `NestedUnidimensionalPointSet` that inherits from `UnidimensionalPointSet`. This derived class will add the levels as an abstract concept. The `_create_points` method should then be required to also handle the level assignment. If we go with this option, it would be better for the `_create_points` method to set `self._points` (and `self._levels` or whatever it's called for this class) directly rather than return by value.
3. Allow classes to add an additional `levels` attribute themselves, and whoever consumes the class will use duck typing to accept anything with `points` and `levels.

I'm not so worried about `MultidimensionalPointSet` because these objects can decide whether they want to use the `levels` or not.

Reply:
I've been hesitant to add level information to the UnidimensionalPointSets and stuck with just number of points at first because if the nested point sets are dependent on the number of levels to determine how many points to generate, then that makes it more cumbersome to use growth schemes that aren't the default. Slow growth and every other growth scheme would need its own class for every unique set of points. I was considering having the levels handled by the MultidimensionalPointSets or by some other object that would exclusively handle levels.

If NestedUnidimensionalPointSets are expected to handle levels, should there be another variable that contains information about how the set is expected to grow each level? The paper I linked in the surrogate models slack channel gives the order function for a ton of growth schemes, which gives exactly how many total points the levels have accumulated. An `order` parameter that's either a callable or a list could be taken with the `max_level` parameter to determine what points are needed and what the list of level indexes should look like. Alternatively, `order` could be a variable that allows the user to select from a predetermined set of allowed order functions, which is what the sgmga code does.

Reply:
To me, how the levels for points are determined is a detail that deriving classes can figure out. This could be achieved in several ways, so I think for now, we just need to decide whether the base unidimensional class has the concept of a level, or if that only belongs in the nested version. Once we make that decision, we would need to survey all the ways for assigning points to levels that we reasonably think we need to support and make a decision about how best to do that.

Reply:
I think that levels should only be a concept for the point sets that the smolyak multidim point set is expected to be able to use. This excludes `ClenshawCurtisPointSet` and `TrigonometricPointSet`, so the base class shouldn't have a concept of levels.

I propose a base class called `LevelUnidimensionalPointSet` or `TieredUnidimensionalPointSet`. (Not `NestedUnidimensionalPointSet` because while the point sets we're using for smolyak are nested, the sets it can use don't necessarily have to be nested.) It will be a child of `UnidimensionalPointSet` and parent to all classes that use levels and that the smolyak grid generator will be allowed to use. 

It will take a parameter `max_level`, and it will have additional read-only properties `levels` and `growth_order`. The `levels` property is identical to the `levels` in the `NestedBasisFunctionSet`, and the base class will have a method for calculating it from `growth_order` and `max_level`. The `growth_order` is an abstract property like `domain` and it is either a callable or a list. It outputs the expected accumulated number of points at each level. Children of this class are expected to give concrete versions of `domain`, `growth_order`, and `_create_points`.

(The reason that `growth_order` is allowed to be callable or a list is to accommodate finite point sets.)

`NestedClenshawCurtisPointSet` would set the `growth_order` to be `lambda x : 1 if x==0 else 2**x + 1`. Replacing its code for generating points with the code that I sent in slack yesterday, I would be able to derive children of `NestedClenshawCurtisPointSet` for slow growth, moderate growth, and linear 2 growth that only need to override `growth_order`.

Reply:
> Doing this will inherently make things nested though, won't it? Since it would not be possible for a point to belong to multiple levels, or to have points added in a non-nested fashion.

In a fully nested rule, the `growth_order` gives the accumulated unique points for a given and all its previous levels, so total new points added at level L is growth_rule(L) - growth_rule(L-1). For a weakly nested (0.0 recurs) or nonnested rule (nothing recurs) this isn't the case, but if a deriving class is nonnested or weakly nested, it can override the method for generating levels.

For nonnested rules it appears to work such that 

`new_points_added(L) = growth_rule(L) if growth_rule(L) - growth_rule(L - 1) != 0 else 0`

and weakly nested rules do this 

`new_points_added(L) = growth_rule(L)-1 if growth_rule(L) - growth_rule(L - 1) != 0 else 0`

For very weakly nested rules, it depends on what levels 0.0 recurs.

The `growth_order` given a max level can easily calculate how many total points are needed, and the number of points can be sent to `_create_points` to generate that many points.

> I also propose that the `_create_points` method be renamed to `_create` in the `UnidimensionalPointSet` class. For normal point sets, it is expected to assign the value of `self._points`. For your level class, it should assign values of both `self._points` and `self._levels`. The `points` (and `levels`) properties should then invoke this method when an internal cache variable has been invalidated.

The internal cache and having the abstract method `_create` for assigning all relevant variables is fine, but I still feel that `points` and `levels` should be handled separately if possible, and not in the same for loop like the old code. Forcing them to both be computed in the same subroutine means that any time we want to use different points or a different growth scheme the entire method for generating both has to be rewritten. Doing them separately isn't going to result in massive runtime penalties. TieredUnidimensionalPointSet can have a defined `_create` method that calls methods `_create_points` and `_create_levels`. Having the `growth_order` property makes creating both as well as the process of generating new classes based off old ones easier.

Reply:
> Forcing them to both be computed in the same subroutine means that any time we want to use different points or a different growth scheme the entire method for generating both has to be rewritten.

This is not necessarily the case. There are ways to package the code for assigning the levels in a reusable way, but we need to talk about what cases we want to enable first.

> Having the growth_order property makes creating both as well as the process of generating new classes based off old ones easier.

I don't think this is necessarily the case either.